### DISCUSSION TITLE
Navigation Refresh/Cache Issue

### DISCUSSION THREAD

**Discussion by DrastiGohel** [serenity-is/Serenity#6768] [Help]
Hello All,

I am facing Navigation refresh issue after assigning roles.

Scenario:

User do not have role permission for a particular section, if a role is assigned to a User, it takes a while for a menu item to get visible in incognito Navigation.

What is needed:

Once the role is assigned the menu should be available to a user immediately in incognito as well as normal log-in.

Is there any code piece to add for that? 

Below is the code for NavigationModelFactory:

```
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Http.Extensions;
using Serenity;
using Serenity.Abstractions;
using Serenity.Navigation;
using Serenity.Web;
using ServicesProject.Navigation;
using ServicesProject.Web.Helpers;
using System;

namespace ServicesProject.Common.Navigation
{
    public class NavigationModelFactory : INavigationModelFactory
    {
        private readonly ITwoLevelCache cache;
        private readonly IHttpContextAccessor httpContextAccessor;
        private readonly ITextLocalizer localizer;
        private readonly IPermissionService permissions;
        private readonly IServiceProvider serviceProvider;
        private readonly ITypeSource typeSource;
        private readonly IUserAccessor userAccessor;

        public NavigationModelFactory(
            ITwoLevelCache cache,
            IHttpContextAccessor httpContextAccessor,
            ITextLocalizer localizer,
            IPermissionService permissions,
            IServiceProvider serviceProvider,
            ITypeSource typeSource,
            IUserAccessor userAccessor)
        {
            this.httpContextAccessor = httpContextAccessor ?? throw new ArgumentNullException(nameof(httpContextAccessor));
            this.localizer = localizer ?? throw new ArgumentNullException(nameof(localizer));
            this.cache = cache ?? throw new ArgumentNullException(nameof(cache));
            this.permissions = permissions ?? throw new ArgumentNullException(nameof(permissions));
            this.serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
            this.typeSource = typeSource ?? throw new ArgumentNullException(nameof(typeSource));
            this.userAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
        }

        private string ToAbsolute(string url)
        {
            return url != null && url.StartsWith("~/", StringComparison.Ordinal) ?
                VirtualPathUtility.ToAbsolute(httpContextAccessor.HttpContext, url) : url;
        }

        private static string NormalizeUrl(string url)
        {
            if (url == null)
                return null;

            if (url.Length > 0 && url[^1] == '?')
                url = url[0..^1];

            if (url.IndexOf('?', StringComparison.Ordinal) < 0 &&
                !url.EndsWith("/", StringComparison.Ordinal))
                url += "/";

            return url;
        }

        public INavigationModel Create()
        {
            var model = new NavigationModel
            {
                Items = cache.GetLocalStoreOnly("LeftNavigationModel:NavigationItems:" +
                    (Convert.ToString(userAccessor.User?.GetUserId()) ?? "-1"), 
                    TimeSpan.Zero,
                    Administration.UserPermissionRow.Fields.GenerationKey,
                    () => NavigationHelper.GetNavigationItems(permissions, typeSource, serviceProvider, ToAbsolute))
            };

            CalcActivePath(model);

            return model;
        }

        private void CalcActivePath(NavigationModel model)
        {
            var currentUrl = httpContextAccessor.HttpContext?.Request.GetEncodedPathAndQuery();

            string bestMatch = null;
            int bestMatchLength = 0;
            NavigationItem bestLink = null;

            foreach (var item in model.Items)
                SearchActivePath(item, currentUrl, ref bestMatch, ref bestMatchLength, ref bestLink);

            model.ActiveItem = bestLink;
        }

        private void SearchActivePath(NavigationItem link, string currentUrl,
            ref string bestMatch, ref int bestMatchLength, ref NavigationItem bestLink)
        {
            if (link == null)
                throw new ArgumentNullException(nameof(link));

            var url = link.Url;

            if (url != null)
            {
                url = ToAbsolute(url);
                url = NormalizeUrl(url);
                if (url.StartsWith(currentUrl, StringComparison.OrdinalIgnoreCase) &&
                    string.Compare(NormalizeUrl(url.Split('?')[0]),
                    NormalizeUrl(currentUrl.Split('?')[0]), StringComparison.OrdinalIgnoreCase) == 0)
                {
                        if (bestMatchLength == 0 || url.Length < bestMatchLength)
                        {
                            bestMatch = link.FullPath;
                            bestMatchLength = url.Length;
                            bestLink = link;
                        }
                }
            }

            foreach (var child in link.Children)
                SearchActivePath(child, currentUrl, ref bestMatch, ref bestMatchLength, ref bestLink);
        }
    }
}
```

Below is code for NavigationModel:

`using System;
using System.Collections.Generic;
using System.Security.Claims;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.DependencyInjection;
using Serenity.Abstractions;
using Serenity.Navigation;
using Serenity.Web;
using ServicesProject.Administration;

namespace ServicesProject.Navigation
{
    public class NavigationModel : INavigationModel
    {
        private NavigationItem activeItem;
        private List<NavigationItem> activePath;

        public NavigationItem ActiveItem
        {
            get => activeItem;
            set
            {
                activeItem = value;
                activePath = null;
            }
        }

        public List<NavigationItem> Items { get; set; }
        IEnumerable<NavigationItem> INavigationModel.Items => Items;

        IEnumerable<NavigationItem> INavigationModel.ActivePath
        {
            get
            {
                if (activePath == null)
                {
                    var p = new List<NavigationItem>();
                    var current = activeItem;
                    while (current != null)
                    {
                        p.Add(current);
                        current = current.Parent;
                    }
                    activePath = p;
                }

                return activePath;
            }
        }

    }
}`

Serenity Version: 6.4.8

**Comment by VictorTomaili** [comment]
You may need to invalidate menu cache on permission—role update. It should be automatically in default. 

The cache is storing left navigation for infinite time per user, until the generation key is changed. You can prefer to shorten the cache timeout. Or you can manually invalidate the UserRoleRow generation key. 

```
cache.GetLocalStoreOnly("LeftNavigationModel:NavigationItems:" +
                    (Convert.ToString(userAccessor.User?.GetUserId()) ?? "-1"), 
                    TimeSpan.Zero,
                    Administration.UserPermissionRow.Fields.GenerationKey,
```

**Reply by rhemasystems** [reply]
I have this already on my latest Serenity by default, but still experiencing the same issue unless I restart the IIS @VictorTomaili

**Reply by reach2rv** [reply]
That code stores cache for indefinite time until generation key is changed. Change timespan or invalidate cache. You can also remove specific cache using key.

**Reply by flashyjunior** [reply]
@reach2rv, can you give me a code snippet to achieve this please, because i am already doing that as seen below

<img width="759" height="511" alt="image" src="https://github.com/user-attachments/assets/663a96fe-58b5-4d41-870b-d6dc388ee911" />

<img width="672" height="446" alt="image" src="https://github.com/user-attachments/assets/9cc82d8e-b09c-4179-add7-0b5b4875e3f2" />
